packages/contracts/test/registry/ens-subdomain-registry.ts

```diff
@@ -0,0 +1,445 @@
+    DUMMY_METADATA,
+    await owner.getAddress(),
+    ethers.constants.AddressZero
+  );
+
+  // Deploy the registrar
+  let registrar = await ENSSubdomainRegistrar.deploy();
+
+  return [ens, resolver, dao, registrar];
+}
+
+describe('ENSSubdomainRegistrar', function () {
+  let signers: SignerWithAddress[];
+  let managingDao: DAO;
+  let ens: ENSRegistry;
+  let resolver: PublicResolver;
+  let registrar: ENSSubdomainRegistrar;
+
+  // A Helper function to register a subdomain under parent domain
+  async function registerSubdomainHelper(
+    subdomain: string,
+    domain: string,
+    domainOwner: SignerWithAddress,
+    subdomainOwnerAddress: string
+  ) {
+    let fullDomain: string;
+    if (domain === '') {
+      fullDomain = subdomain;
+    } else {
+      fullDomain = subdomain + '.' + domain;
+    }
+
+    let tx = await ens
+      .connect(domainOwner)
+      .setSubnodeRecord(
+        ensDomainHash(domain),
+        ensLabelHash(subdomain),
+        subdomainOwnerAddress,
+        resolver.address,
+        0
+      );
+    await tx.wait();
+
+    // Verify that the subdomain is owned by the correct address
+    expect(await ens.owner(ensDomainHash(fullDomain))).to.equal(
+      subdomainOwnerAddress
+    );
+    // Verify that that the subdomain's resolver address is set correctly
+    expect(await ens.resolver(ensDomainHash(fullDomain))).to.equal(
+      resolver.address
+    );
+  }
+
+  beforeEach(async () => {
+    signers = await ethers.getSigners();
+  });
+
+  describe('After deployment', () => {
+    beforeEach(async () => {
+      [ens, resolver, managingDao, registrar] = await setupENS(signers[0]);
+    });
+
+    it('unregistered domains are owned by the zero address on ENS', async () => {
+      expect(await ens.owner(ensDomainHash('test'))).to.equal(
+        ethers.constants.AddressZero
+      );
+    });
+
+    it('unregistered domains resolve to the zero address on ENS', async () => {
+      expect(await resolver['addr(bytes32)'](ensDomainHash('test'))).to.equal(
+        ethers.constants.AddressZero
+      );
+    });
+
+    it('reverts during intialization because the registrar is neither the domain node owner nor an approved operator of him/her', async () => {
+      await expect(
+        registrar
+          .connect(signers[1])
+          .initialize(managingDao.address, ens.address, ensDomainHash('test'))
+      ).to.be.revertedWith(
+        customError(
+          'UnauthorizedRegistrar',
+          ethers.constants.AddressZero,
+          registrar.address
+        )
+      );
+
+      // This is also the case for signers[0] who owns the ENS registries
+      await expect(
+        registrar
+          .connect(signers[0])
+          .initialize(managingDao.address, ens.address, ensDomainHash('test'))
+      ).to.be.revertedWith(
+        customError(
+          'UnauthorizedRegistrar',
+          ethers.constants.AddressZero,
+          registrar.address
+        )
+      );
+    });
```

Comment:
This is a follow-up of that:
> Could we test the 2 conditions separately? (owner / authorized)
> Also, shouldn't we be trying to connect using another random signer?

Given this test description: `reverts during intialization because the registrar is neither the domain node owner nor an approved operator of him/her`.

I would:
1. Set up scenario 1
  - Registrar is the domain owner but NOT approved
  - Check that it works
2. Set up scenario 2
  - Registrar is NOT the domain owner but it is approved
  - Check that it works
3. Set up scenario 3
  - Registrar is the domain owner AND also it is approved
  - Check that it works
4. Set up scenario 4
  - Registrar is NOT the domain owner and is NOT approved
  - Check that it does NOT work (revert)
5. Set up scenario 5
  - Random signer with no permissions at all
  - Check that it does NOT work (revert)

Reply:
Original: https://github.com/aragon/core/pull/51#discussion_r895831454

Reply:
I tried to clarify this by rearranging / renaming the tests in the latest commits.